Implementing GAN compute utils

szmazurek opened the pull request:
Fixes #813

## Proposed Changes
- added all compute utilities required to train the network
- they are kept separate, as the GAN folder will serve as closed testing playground for the synthesis solutions

## Checklist
- [X] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [X] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [X] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [X] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [X] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].

szmazurek committed:
Implementing compute utils

- added all compute utilities required to train the network
- they are kept separate, as the GAN folder will serve as closed testing
  playground for the synthesis solutions

sarthakpati commented on GANDLF/GAN/compute/forward_pass.py:
```diff
@@ -0,0 +1,395 @@
+            patch_overlap=params["inference_mechanism"]["patch_overlap"],
+        )
+        patch_loader = torch.utils.data.DataLoader(grid_sampler, batch_size=1)
+        # aggregator = torchio.inference.GridAggregator(
```
Put a comment (and related documentation) that the synthesis framework only supports full image validation.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/forward_pass.py:
```diff
@@ -0,0 +1,395 @@
+        ]:
+            raise NotImplementedError(
+                "Reduce on plateau scheduler not implemented for GAN"
+            )
```
Replace with `assert` - improves code coverage, and it is consistent with what is happening in the rest of the code.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/forward_pass.py:
```diff
@@ -0,0 +1,395 @@
+        ]:
+            raise NotImplementedError(
+                "Reduce on plateau scheduler not implemented for GAN"
+            )
```
Replace with `assert` - improves code coverage, and it is consistent with what is happening in the rest of the code.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/inference_loop.py:
```diff
@@ -0,0 +1,206 @@
+            "The dataframe was passed, but it's usage is not yet implemented in inference. Ignoring it."
+        )
+
+    if parameters["model"]["type"].lower() == "openvino":
+        raise NotImplementedError("OpenVINO inference is not yet implemented")
```
Replace with `assert` - improves code coverage, and it is consistent with what is happening in the rest of the code.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/inference_loop.py:
```diff
@@ -0,0 +1,206 @@
+        model.load_state_dict(main_dict["model_state_dict"])
+        parameters["previous_parameters"] = main_dict.get("parameters", None)
+        model.eval()
+    elif parameters["model"]["type"].lower() == "openvino":
+        raise NotImplementedError("OpenVINO inference is not yet implemented")
```
Replace with `assert` - improves code coverage, and it is consistent with what is happening in the rest of the code.

szmazurek replied:
Done, plus no idea why I repeated that condition lol

sarthakpati commented on GANDLF/GAN/compute/generic.py:
```diff
@@ -0,0 +1,199 @@
+        batch_size (int): The batch size.
+        latent_vector_size (int): The latent vector size.
+        dimension (int): The dimension of the images in a given problem.
+    can be 2 for 2D or 3 for 3D.
```
```suggestion
        dimension (int): The dimension of the images in a given problem. Can be `2` for 2D or `3` for 3D.
```

sarthakpati reviewed (changes requested):
Requesting some minor changes. Additionally, these:

- Rename `GAN` to `synthesis`, because we might want to add in diffusion models at some point of time as well.
- Update the function documentation style according to how the rest of it is structured. See [this](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/compute/forward_pass.py#L35-L48) for reference.

szmazurek committed:
Update GANDLF/GAN/compute/generic.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Fixing the comment on image loading

szmazurek committed:
Update forward_pass.py

Convert error raise into the assert for code coverage and coherence.

szmazurek committed:
Hotfix

- assertion correction in forward_pass.py
- assertion created in inference_loop.py

szmazurek commented:
> Requesting some minor changes. Additionally, these:
> 
> * Rename `GAN` to `synthesis`, because we might want to add in diffusion models at some point of time as well.
> * Update the function documentation style according to how the rest of it is structured. See [this](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/compute/forward_pass.py#L35-L48) for reference.

Regarding first one - noted that, will do separate pr for that when the rest will be merged
Regarding second one - 5 mins and push

szmazurek committed:
Fixing typing and docstring:

- made function description compatibile with the one already present in
  the main GANDLF code
- updated the function when needed to incorporate changes

sarthakpati commented on GANDLF/GAN/compute/forward_pass.py:
```diff
@@ -0,0 +1,387 @@
+        )
+    pathlib.Path(current_output_dir).mkdir(parents=True, exist_ok=True)
+
+    # I really do not get it
```
Huh? 😆

szmazurek replied:
I will think about it, there can be more of such desperation artifacts when I was trying to solve it :p anyway, will come back to you with this one

sarthakpati replied:
Okay - but the comment needs to make some sense. 🤣

szmazurek replied:
Ok, I know what I meant, added that to the comment :P

szmazurek replied:
Solved

sarthakpati commented on GANDLF/GAN/compute/generic.py:
```diff
@@ -0,0 +1,198 @@
+def generate_latent_vector(
+    batch_size: int, latent_vector_size: int, dimension: int, device: str
+) -> torch.Tensor:
+    """Creates a latent vector of given size and adjusts the dimensions
```
Update comment style (for reference, see the documentation for `create_pytorch_objects_gan`).

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/generic.py:
```diff
@@ -0,0 +1,198 @@
+    seed: int,
+) -> torch.Tensor:
+    """
+    Function to get the fixed latent vector for inference or validation.
```
Update comment style (for reference, see the documentation for `create_pytorch_objects_gan`).

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/generic.py:
```diff
@@ -0,0 +1,198 @@
+
+
+def create_pytorch_objects_gan(
+    parameters, train_csv=None, val_csv=None, device="cpu"
```
Add typing information.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/inference_loop.py:
```diff
@@ -0,0 +1,206 @@
+
+## TODO move this function into some utils file
+def norm_range(t: torch.Tensor):
+    """Normalizes the input tensor to be in the range [0, 1]. Operation is
```
Update comment style (for reference, see the documentation for `inference_loop_gan`).

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/inference_loop.py:
```diff
@@ -0,0 +1,206 @@
+    """
+
+    def norm_ip(img: torch.Tensor, low: float, high: float):
+        """Utility function to normalize the input image, the same as in
```
Update comment style (for reference, see the documentation for `inference_loop_gan`).

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/inference_loop.py:
```diff
@@ -0,0 +1,206 @@
+    )  # how many iterations to run
+    remaining_samples = (
+        n_generated_samples % batch_size
+    )  # remaining samples for last iteration
```
Run `black` on this again.
```suggestion
    # how many iterations to run
    n_iterations = (
        n_generated_samples // batch_size
    )  
    # remaining samples for last iteration
    remaining_samples = (
        n_generated_samples % batch_size
    )  
```

sarthakpati commented on GANDLF/GAN/compute/loss_and_metric.py:
```diff
@@ -0,0 +1,92 @@
+
+def get_metric_output(metric_function, predicted, ground_truth, params):
+    """
+    This function computes the output of a metric function.
```
Update comment style.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/loss_and_metric.py:
```diff
@@ -0,0 +1,92 @@
+
+def get_loss_gans(predictions: Tensor, labels: Tensor, params: dict) -> Tensor:
+    """
+    Compute the loss value for adversatial generative networks.
+    Args:
+        predictions (Tensor): The predicted output from the model.
+        labels (Tensor): The ground truth label.
+        params (dict): The parameters passed by the user yaml.
+    Returns:
+        loss (Tensor): The computed loss from the label and the prediction.
```
```suggestion
    Compute the loss value for adversarial generative networks.

    Args:
        predictions (Tensor): The predicted output from the model.
        labels (Tensor): The ground truth label.
        params (dict): The parameters passed by the user yaml.

    Returns:
        loss (Tensor): The computed loss from the label and the prediction.
```

szmazurek replied:
done, will be commited

sarthakpati commented on GANDLF/GAN/compute/step.py:
```diff
@@ -0,0 +1,100 @@
+    Union[torch.Tensor, None],
+]:
+    """
+    Function that steps the GAN model for a single batch
```
Update comment style.

sarthakpati commented on GANDLF/GAN/compute/training_loop.py:
```diff
@@ -0,0 +1,710 @@
+    model, train_dataloader, optimizer_g, optimizer_d, params
+):
+    """
+    Function to train a GAN network for a single epoch.
```
Update comment style.

szmazurek replied:
Done

sarthakpati commented on GANDLF/GAN/compute/training_loop.py:
```diff
@@ -0,0 +1,710 @@
+        tqdm(train_dataloader, desc="Looping over training data")
+    ):
+        #### DISCRIMINATOR STEP WITH ALL REAL LABELS ####
+        optimizer_d.zero_grad()
```
Wondering if it would make sense to split all this code into one or more `step` functions for maintainability... Thoughts?

szmazurek replied:
That is actually right - do you think to that I should go for it right now or we will split it in the future as a separate pr? Maintainability-wise, this is definetly better option.

sarthakpati replied:
I would suggest to do it now itself so that we are not scrambling with additional stuff later...

szmazurek replied:
I am running a training to see if it did not break anything, I have moved them to separate functions, as well as backward pass itself. Will push shortly if the test is ok.

szmazurek replied:
Done, changes in all files are purely code formatting related. The changes are in training_loop.py file, GAN steps extracted and refactored.

sarthakpati reviewed (commented):
Minor stylistic changes and one philosophical comment.

szmazurek committed:
Update forward_pass.py

- remove unclear comment
- remove other unecessary comments

szmazurek committed:
Docs and typing updateds

- updated comment styles in functions generating latent vectors
- updated typing in function to create pytorch objects

szmazurek committed:
Docstring updates for utility functions

- updated docstrings for normalizing functions

szmazurek committed:
Update GANDLF/GAN/compute/inference_loop.py

Styling updates

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Updating typing and docstrings

- fixed docstrings layout
- added typing annotations in get_metric_output function
- modified and extended typing in get_loss_and_metrics_gans function

szmazurek committed:
Updating typing and docstrings:

- updated the docstring and types for step_gan function

szmazurek committed:
Updated typing and docstring in train_network_gan

szmazurek committed:
Refactoring

- extracted the discirminator and generator steps into two separate
  functions
- separated the backward_pass function

szmazurek committed:
Code formatting and refactoring

- ran black formatting with new settings
- fixed small bug in the newly refactored functions

sarthakpati commented on GANDLF/GAN/compute/training_loop.py:
```diff
@@ -193,9 +193,11 @@ def generator_step(
     label_fake = torch.full(
         size=(current_batch_size,),
-        fill_value=0,
+        fill_value=1,
```
Why is the fill value changing from `0` to `1`?

szmazurek replied:
in previous pr I set it incorrectly to 0, it should be 1

sarthakpati replied:
Understood.

sarthakpati reviewed (commented):
Small question

sarthakpati reviewed (approved):
LGTM
